Add process to rearrange counts and logcounts in processed AnnData objects

allyhawkins opened the pull request:
Closes #429 and #436

This PR adds a new process to the workflow for exporting AnnData to rearrange the processed AnnData objects only. For each processed object, we move the `X` assay (counts) to `raw.X` and then move the `logcounts` layer to `X`. 

- This approach means that every AnnData object is saved in the same way when objects are converted from SCE to AnnData. Then, we read in the processed objects only and check if `logcounts` is present in the `layers`. If that's true, then we rearrange. 
- I decided to take the same approach for both RNA and feature processed objects. They are passed in as a list of files and then we check for `logcounts` in each file before doing any rearranging. 
- Should we add some sort of note in the metadata about counts being moved? I did change the `X_name` though so I think that should be sufficient?

allyhawkins commented:
This is failing checks right now because it can't find `anndata` even though it should be in the image...

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -44,11 +61,21 @@ workflow sce_to_anndata{
       // export each anndata file
       export_anndata(sce_ch)
 
+      processed_anndata_ch = export_anndata.out
+        .filter{ it[2] == "processed"}
```
rather than filtering in two places, use `branch` to centralize the logic
```suggestion
     anndata_ch = export_anndata.out
        .branch{
          processed: it[2] == "processed"
          other: true
        }
```

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -44,11 +61,21 @@ workflow sce_to_anndata{
+
       // combine all anndata files by library id
       // creates anndata channel with [library_id, unfiltered, filtered, processed]
       anndata_ch = export_anndata.out
```
I can't make the full suggestion here, but replacing everything up to the `.map`
```suggestion
      anndata_ch = anndata_ch.other.mix(move_normalized_counts.out)
```

jashapiro commented on modules/export-anndata.nf:
```diff
@@ -44,11 +61,21 @@ workflow sce_to_anndata{
+          meta.library_id,
           meta,
           hdf5_files
         )}
```
I can't quite comment on it below, but we might want a better comment on what is going on in the `.map` below (I think it could probably be cleaned up with a more explicit map statement.

jashapiro reviewed (commented):
I think the main issue here is a `stub` field not being present... I don't think it uses the docker image for stubs. I'm assuming you ran this with real data and it worked?

jashapiro commented on bin/move_counts_anndata.py:
```diff
@@ -0,0 +1,46 @@
+# compile extension regex
+file_ext = re.compile(r"\.hdf5$|.h5$", re.IGNORECASE)
+
+for file in input_files:
```
I'm going to suggest simplifying the script to take only one file at a time and then looping through the files in the outer process to save nesting splitting, etc.

(I'd also rename the argument to `--anndata_file` or something like that as this is both input and output)

Nextflow is pretty smart about passing in lists of files, so you can do something like (note that the shell variable has to be escaped)
```
script:
  """
  for file in ${processed_hdf5}; do
    move_counts_anndata.py --anndata_file \${file}
  done
  """
```

allyhawkins commented:
Thanks for looking at this @jashapiro. I added the stub and also changed the script to only take in one input file. I tested the new changes and everything worked. This is ready for another look.

allyhawkins commented:
I added formatting to this with `black` and then also added the `schema_version` to the AnnData object before exporting. 
Other than the title, that's the only thing that needs to be added to `uns` so thought this was a good time to add it: https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#uns-dataset-metadata

jashapiro commented on bin/move_counts_anndata.py:
```diff
@@ -0,0 +1,50 @@
+# if logcounts is present
+if "logcounts" in object.layers.keys():
+    # move X to raw.X by creating the raw object
+    object.raw = object
```
Does copying this way end up copying _everything_ to `object.raw`? It seems like that would create a lot of extra data. (i.e. is there now `object.raw.layers["logcounts"]`?) Looking at [the docs](https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.raw.html), it seems like no, but I just want to confirm.

allyhawkins replied:
No, it looks like it only stores `X` and `var`.

jashapiro reviewed (approved):
LGTM! I had a little question/fear about how `.raw` is created, but I'm pretty sure the answer is what it should be. It just looks a bit strange which is no fault of yours!
